solidity/test/tbtc/Vault.test.ts

```diff
@@ -0,0 +1,442 @@
+import { ethers, getUnnamedAccounts, helpers, waffle } from "hardhat"
+import { expect } from "chai"
+
+import { ContractTransaction } from "ethers"
+import type { Bank, TBTC, Vault } from "../../typechain"
+
+const { to1e18 } = helpers.number
+const { createSnapshot, restoreSnapshot } = helpers.snapshot
+
+const ZERO_ADDRESS = ethers.constants.AddressZero
+
+const fixture = async () => {
+  const [deployer, bridge] = await ethers.getSigners()
+
+  const Bank = await ethers.getContractFactory("Bank")
+  const bank = await Bank.deploy()
+  await bank.deployed()
+
+  await bank.connect(deployer).updateBridge(bridge.address)
+
+  const TBTC = await ethers.getContractFactory("TBTC")
+  const tbtc = await TBTC.deploy()
+  await tbtc.deployed()
+
+  const Vault = await ethers.getContractFactory("Vault")
+  const vault = await Vault.deploy(bank.address, tbtc.address)
+  await vault.deployed()
+
+  await tbtc.connect(deployer).transferOwnership(vault.address)
+
+  return {
+    bridge,
+    bank,
+    vault,
+    tbtc,
+  }
+}
+
+describe("Vault", () => {
+  let bridge: SignerWithAddress
+  let bank: Bank
+  let vault: Vault
+  let tbtc: TBTC
+
+  const initialBalance = to1e18(100)
+
+  let account1: SignerWithAddress
+  let account2: SignerWithAddress
+
+  before(async () => {
+    // eslint-disable-next-line @typescript-eslint/no-extra-semi
+    ;({ bridge, bank, vault, tbtc } = await waffle.loadFixture(fixture))
+
+    const accounts = await getUnnamedAccounts()
+    account1 = await ethers.getSigner(accounts[0])
+    account2 = await ethers.getSigner(accounts[1])
+
+    await bank.connect(bridge).increaseBalance(account1.address, initialBalance)
+    await bank.connect(bridge).increaseBalance(account2.address, initialBalance)
+
+    await bank.connect(account1).approveBalance(vault.address, initialBalance)
+    await bank.connect(account2).approveBalance(vault.address, initialBalance)
+  })
+
+  describe("constructor", () => {
+    context("when called with a 0-address bank", () => {
+      it("should revert", async () => {
+        const Vault = await ethers.getContractFactory("Vault")
+        await expect(
+          Vault.deploy(ZERO_ADDRESS, tbtc.address)
+        ).to.be.revertedWith("Bank can not be the zero address")
+      })
+    })
+
+    context("when called with a 0-address TBTC token", () => {
+      it("should revert", async () => {
+        const Vault = await ethers.getContractFactory("Vault")
+        await expect(
+          Vault.deploy(bank.address, ZERO_ADDRESS)
+        ).to.be.revertedWith("TBTC token can not be the zero address")
+      })
+    })
+
+    context("when called with correct parameters", () => {
+      it("should set the Bank field", async () => {
+        expect(await vault.bank()).to.equal(bank.address)
+      })
+
+      it("should set the TBTC token field", async () => {
+        expect(await vault.tbtcToken()).to.equal(tbtc.address)
+      })
+    })
+  })
+
+  describe("mint", () => {
```

Comment:
Would it make sense to add a case when one minter uses multiple vaults to mint different assets and assert the bank correctly updates its balance?

Reply:
Isn't it something already covered by `Bank.test.ts`? From the Bank's perspective, it does not matter if the balance owner is EOA or smart contract and it is just transferring balances the owner gave an allowance for.

Reply:
Yep right, `Bank.test.ts` covers that. I was just thinking about making it explicit here but right, that feels redundant.